Handle mipmaps and VRAM compression for glTF binary images

fire opened the pull request:
Provide three options:

[ Ignore and Warn (default) | Optimize Loading Embedded as Basisu | Extract ]

----

Fixes: https://github.com/godotengine/godot-proposals/issues/4271

Relates: https://github.com/godotengine/godot/issues/62436

Relates: https://github.com/godotengine/godot/issues/62295

----

Part of the V-Sekai group. https://github.com/v-sekai

fire commented:
See also #59540.

lyuma commented:
@fire Do you have some data on the performance problems?

I think this is what we were trying to figure out when we were discussing the viability of Basis a few months ago.

Is the problem with import / compression (okay) or runtime / decompression / conversion to native VRAM compression (bad, affects game loading time)

fire commented:
I think we need time limits for the import. I tested using a sample non commercial licensed avatar like https://github.com/Miraikomachi/MiraikomachiVRM/blob/master/Miraikomachi.vrm and it was very slow with our vrm addon.

fire commented:
Original use case:

1. Have a single packed godot scene to store the entire level and expecting the level's texture images to work on android, mac and pc.
2. Use basis u and portable compressed texture
3. and the images will be portable compressed texture
4. if one disables uastc, it will be bad quality

fire committed:
Handle gltf binary

[ Ignore and Warn | Extract Textures (default) | Optimize Loading Embedded as Basisu ]

Enable compressed mip maps from Basis Universal for faster compressions.

Increase the quality of Basis to avoid corruption.

To keep compatibility use the first mip of the previous internal Godot format.

Because texture names may have invalid filename characters, adds String::validate_filename to sanitize filenames for import pipeline use.

reduz commented:
I think we need a better way to tell users they exported their GLTF with built-in textures and by default Godot will take no action with them (we could make it optional what to do per scene in the GLTF import options).

fire commented:
Discussed with @reduz 

We concluded with this design.

[ Ignore and Warn (default) | Optimize Loading Embedded as Basisu etc1s via https://github.com/godotengine/godot/pull/59540 | Extract ]

fire commented:
We discussed avoiding the folder.

Do filename.gltf -> filename[minus .gltf].texturename.png

I'll try to provide a pull request where the GLTF2 can compress to BasisU (etc1s) and extract to the folder using the above convention.

fire commented:
@reduz 

I have tested all three setups:

1. Ignore and Warn
2. Optimize Loading Embedded as Basisu etc1s via https://github.com/godotengine/godot/pull/59540
3. Extract
4. Current behaviour (default) 

This was done by manually applying the flags.

```c++
Node *EditorSceneFormatImporterGLTF::import_scene(const String &p_path, uint32_t p_flags,
		const HashMap<StringName, Variant> &p_options, int p_bake_fps,
		List<String> *r_missing_deps, Error *r_err) {
	Ref<GLTFDocument> doc;
	doc.instantiate();
	Ref<GLTFState> state;
	state.instantiate();
	p_flags |= EditorSceneFormatImporter::IMPORT_EXTRACT_TEXTURES; // TODO: Testing REMOVE fire 2022-08-07 // **HERE**
	Error err = doc->append_from_file(p_path, state, p_flags, p_bake_fps);
	if (err != OK) {
		if (r_err) {
			*r_err = err;
		}
		return nullptr;
	}
	return doc->generate_scene(state, p_bake_fps);
}
```

fire commented on modules/gltf/gltf_state.cpp:
```diff
@@ -82,12 +82,20 @@ void GLTFState::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("set_animations", "animations"), &GLTFState::set_animations);
 	ClassDB::bind_method(D_METHOD("get_scene_node", "idx"), &GLTFState::get_scene_node);
-
+	ClassDB::bind_method(D_METHOD("get_discard_textures"), &GLTFState::get_discard_textures);
+	ClassDB::bind_method(D_METHOD("set_discard_textures", "discard_textures"), &GLTFState::set_discard_textures);
+	ClassDB::bind_method(D_METHOD("get_embed_as_basisu"), &GLTFState::get_embed_as_basisu);
+	ClassDB::bind_method(D_METHOD("set_embed_as_basisu", "embed_as_basisu"), &GLTFState::set_embed_as_basisu);
+	ClassDB::bind_method(D_METHOD("get_extract_textures"), &GLTFState::get_extract_textures);
+	ClassDB::bind_method(D_METHOD("set_extract_textures", "extract_textures"), &GLTFState::set_extract_textures);
```
Should be an enum.

fire commented:
Pending combining discard_textures, embed_as_basisu and extract_texture into one enum.

reduz commented on editor/import/resource_importer_scene.h:
```diff
@@ -67,6 +67,9 @@ class EditorSceneFormatImporter : public RefCounted {
 		IMPORT_GENERATE_TANGENT_ARRAYS = 8,
 		IMPORT_USE_NAMED_SKIN_BINDS = 16,
 		IMPORT_DISCARD_MESHES_AND_MATERIALS = 32, //used for optimizing animation import
+		IMPORT_DISCARD_TEXTURES = 64,
```
I added the possibility for `EditorSceneFormatImporter` to supply import options, so please avoid this hack and use this.

fire replied:
EditorSceneFormatImporter uses import options, which then feed into the gltf document options, will need to change the  gltf document api.

reduz commented on modules/gltf/gltf_document.cpp:
```diff
@@ -3199,6 +3207,43 @@ Error GLTFDocument::_parse_images(Ref<GLTFState> state, const String &p_base_pat
+			state->images.push_back(Ref<Texture2D>());
+			continue;
+		}
+		if (state->extract_textures) {
```
This will probably need changes to notify the importer that a texture was saved, but I guess its out of the scope of this PR.

Also, when just opening the scene (Before import options are passed) I think these flags need to be ignored since they are for previsualization.

fire commented:
Working on this as soon as I can.

fire commented:
@reduz I got stuck reimporting the textures when extracted.

fire commented on modules/gltf/gltf_state.h:
```diff
@@ -31,6 +31,7 @@
 #ifndef GLTF_STATE_H
 #define GLTF_STATE_H
 
+#include "core/object/class_db.h"
```
Removing.

lyuma commented on modules/basis_universal/register_types.cpp:
```diff
@@ -52,44 +52,50 @@ enum BasisDecompressFormat {
+		basisu::vector<basisu::image> source_images;
+		for (int32_t mipmap_i = 1; mipmap_i < image->get_mipmap_count(); mipmap_i++) {
+			Ref<Image> mip = image->get_image_from_mipmap(mipmap_i);
+			basisu::image buimg(mip->get_width(), mip->get_height());
```
Please test with non-power of two and non-multiple of 4 to see if anything needs to be changed.

While LOD 0 must be multiple of 4, it is legal for mipmaps to be non-multiple of 4. Make sure basis supports non-padded non-multiple of 4 mips properly. it may need to be padded depending on the API

fire replied:
The current code forces power of two for all basisu textures.

lyuma commented on modules/gltf/gltf_document.cpp:
```diff
@@ -3258,13 +3316,26 @@ GLTFTextureIndex GLTFDocument::_set_texture(Ref<GLTFState> state, Ref<Texture2D>
+	if (state->handle_binary_image != GLTFState::GLTF_DOCUMENT_EMBED_AS_BASISU) {
+		return state->images[image];
+	}
+	if (!p_is_normal) {
+		return state->images[image];
+	}
```
This should be cached for both normal and non-normal cases.

If a normal-map is used 100 times or even twice, we cannot afford to make duplicate copies of the texture in this case.

fire replied:
I think a full caching system is material for a different pr, need to make sure it applies to everything and this pr is already dense.

fire replied:
I misread the comment. It now caches.

lyuma commented on modules/gltf/gltf_document.cpp:
```diff
@@ -3258,13 +3316,26 @@ GLTFTextureIndex GLTFDocument::_set_texture(Ref<GLTFState> state, Ref<Texture2D>
 }
 
-Ref<Texture2D> GLTFDocument::_get_texture(Ref<GLTFState> state, const GLTFTextureIndex p_texture) {
+Ref<Texture2D> GLTFDocument::_get_texture(Ref<GLTFState> state, const GLTFTextureIndex p_texture, bool p_is_normal) {
```
Do we also require an argument for p_is_roughness for ORM / roughness maps and passing in the associated normal map?

Godot uses the roughness flag to enable the roughness limiter on that texture channel.

fire replied:
The roughness definition is only known when we have a material.

As far as I know the only important distinctions is the source_color, and the normals image types other than regular images.

lyuma replied:
glTF files have a material. Therefore we should be able to know which channels and textures are used for normals and roughness, same way we know something is a normal map

fire replied:
I'll try to implement this after ResourceLoader::import is done.

fire replied:
I added an enum but I didn't see any cases.

lyuma reviewed (changes requested):
> 3227 file_path image->get_name():

Please sanitize img->get_name(). I don't want bugs or errors caused by gltf image names with slashes (or other illegal characters on windows).

lyuma commented on modules/gltf/gltf_document.cpp:
```diff
@@ -3196,10 +3204,60 @@ Error GLTFDocument::_parse_images(Ref<GLTFState> state, const String &p_base_pat
+				state->images.push_back(Ref<Texture2D>());
+				state->source_images.push_back(Ref<Image>());
+			} else {
+				String file_path = state->get_base_path() + state->filename.get_basename() + "_" + img->get_name() + ".png";
```
sanitize

fire replied:
Can you provide example code?

lyuma replied:
does it also need de-duplication (accounting for non-case-sensitive filesystems) or handling of images which don't have names?

fire replied:
If the buffer view is nameless, it is set to the number.

If the buffer has duplicates.. I don't think that's allowed in gltf.

lyuma replied:
The issue is they may have duplicates after accounting for case-insensitivity in filesystems. This could lead to behavior which breaks only on windows.

lyuma replied:
Of course, now we start getting into other filesystem quirks like unicode normalization such as on APFS on macos: https://twitter.com/Foone/status/1369516978029793283

Honestly, for our case, I'm not as concerned about collisions (duplicate names or case sensitivity or normalization). So let's shelve this whole part of the concern.

Let's only fix cases that are outright illegal (such as double-quotes or pipe characters), which is what that `_make_extname` function does.

fire replied:
Need to check what happens when an illegal name is saved on windows and is renamed.

fire replied:
Added a local solution rather than a global solution to Saver.

lyuma commented on modules/gltf/gltf_document.cpp:
```diff
@@ -3196,10 +3204,60 @@ Error GLTFDocument::_parse_images(Ref<GLTFState> state, const String &p_base_pat
+				state->source_images.push_back(Ref<Image>());
+			} else {
+				String file_path = state->get_base_path() + state->filename.get_basename() + "_" + img->get_name() + ".png";
+				img->save_png(file_path);
+				Ref<Texture2D> saved_image = ResourceLoader::load(file_path, "Texture2D");
```
I wasn't aware that you can load a Texture2D from a newly-written path without going through the importer pipeline (e.g. calling reimport_files() or another API). Does this load the texture uncompressed?

fire replied:
This part is broken. Stuck.

lyuma replied:
reduz noted in RocketChat to look at how the lightmapper does this. Lightmaps use ResourceLoader::import() followed by ResourceLoader::load().

lyuma reviewed (commented):
Overall pretty good. Here are two more issues I found:

fire commented:
`extracted_image_name` is defined as `String extracted_image_name = image_name.get_file().get_basename();`

I'll translate. It means get the file like `blah/blah.png` becomes `blah.png` and then get its basename which is `blah`.

fire commented:
From Lyuma.

> lightmap_gi.cpp, line 155:

		Ref<ConfigFile> config;
		config.instantiate();

		if (FileAccess::exists(texture_path + ".import")) {
			config->load(texture_path + ".import");
		}

		config->set_value("remap", "importer", "2d_array_texture");
		config->set_value("remap", "type", "CompressedTexture2DArray");
		if (!config->has_section_key("params", "compress/mode")) {
			config->set_value("params", "compress/mode", 2); //user may want another compression, so leave it be
		}
		config->set_value("params", "compress/channel_pack", 1);
		config->set_value("params", "mipmaps/generate", false);
		config->set_value("params", "slices/horizontal", 1);
		config->set_value("params", "slices/vertical", texture_slice_count);

		config->save(texture_path + ".import");

		Error err = texture_image->save_exr(texture_path, false);
		ERR_FAIL_COND_V(err, ret);
		ResourceLoader::import(texture_path);
		Ref<TextureLayered> t = ResourceLoader::load(texture_path); //if already loaded, it will be updated on refocus?
		ERR_FAIL_COND_V(t.is_null(), ret);
		ret[i] = t;

ResourceLoader::import

fire commented:
Comment. I removed the texture caching code again.

fire commented:
![image](https://user-images.githubusercontent.com/32321/209578019-c390b391-7712-4146-b6c4-673aebf7ca13.png)

A bug is the portable texture requires knowing about if the texture is a normal map or not but when we go through gltf -> portable compressed 2d texture -> basisu we lose that when compressing basisu as if it was a base color map.

clayjohn commented:
@Fire what remains to be done on this PR before it is ready? It would be nice for this to make it in time for 4.0 so we can support ``.glb`` files better

fire commented:
I believe this was blocked on ASTC support (done by one of my prs), which is blocked on ASTC (reduz expressed interest in doing this) which is blocked on Basisu changes (easy).

fire commented:
Use the image naming at the same level don't use a folder.

fire commented:
I am happy with regards to usability.

[ Extract (default) | Ignore and Warn | Optimize Loading Embedded as Basisu ]

lyuma commented on modules/gltf/gltf_document.cpp:
```diff
@@ -3195,9 +3201,92 @@ Error GLTFDocument::_parse_images(Ref<GLTFState> p_state, const String &p_base_p
+			extracted_image_name = extracted_image_name.replace("@", "_");
+			extracted_image_name = extracted_image_name.replace("`", "_");
+			extracted_image_name = extracted_image_name.replace("|", "_");
+			extracted_image_name = extracted_image_name.replace("=", "_");
```
Added a validate_filename function with fire.

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -4352,9 +4352,22 @@ bool String::is_valid_filename() const {
+
+String String::validate_filename() const {
+	Vector<String> chars = String(invalid_filename_characters).split(" ");
+	String name = this->replace(chars[0], "_");
+	for (int i = 1; i < chars.size(); i++) {
+		name = name.replace(chars[i], "_");
+	}
```
Isn't this clearer?
```suggestion
	String name = this;
	for (int i = 0; i < chars.size(); i++) {
		name = name.replace(chars[i], "_");
	}
```

lyuma replied:
Makes sense. I copied this from `validate_node_name` so I didn't have to think too hard :-p

lyuma replied:
I forgot to put in the strip_edges(), so I changed it to `String name = strip_edges()`

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -4352,9 +4352,22 @@ bool String::is_valid_filename() const {
+static const char *invalid_filename_characters = ": / \\ ? * \" | % < >";
+
+String String::validate_filename() const {
+	Vector<String> chars = String(invalid_filename_characters).split(" ");
```
Could be `LocalVector` maybe.

lyuma replied:
The issue is .split() allocates a Vector. `validate_node_name` does the same. It's called infrequently in imoprt code.
I was originally thinking of doing .replace().replace()...

akien-mga replied:
It can be for a follow-up PR if needed/relevant, but I guess `invalid_filename_characters` could be made a static `LocalVector<CharString>` with all the chars already split, so we don't need to redo this computation for what's going to always be the same every time we validate a filename.

akien-mga commented on core/string/ustring.cpp:
```diff
@@ -4352,9 +4352,22 @@ bool String::is_valid_filename() const {
 		return false;
 	}
 
+	// TODO: Use invalid_filename_characters
 	return !(find(":") != -1 || find("/") != -1 || find("\\") != -1 || find("?") != -1 || find("*") != -1 || find("\"") != -1 || find("|") != -1 || find("%") != -1 || find("<") != -1 || find(">") != -1);
```
Would be good to solve the TODO now, shouldn't be hard.

One solution would be `return this == this->validate_filename();` but it's likely more expensive than checking char by char, and it would be good to keep this as fast as possible.

So I guess (untested):
```cpp
Vector<String> chars = String(invalid_filename_characters).split(" ");
for (const String &char : chars) {
	if (contains(char)) {
		return false;
	}
}
return true;
```

lyuma replied:
okay, done. We do seem to have a unit test for this.

akien-mga commented:
Thanks!
